### PR TITLE
feat: support period vote tx

### LINKED ISSUE #282: add support for PERIOD_VOTE tx type
## Bounty

period vote transaction are used by validators to signal on data availability of 32-block chunks.

## Scope

- enable node to handle [period vote txns](https://github.com/leapdao/leap-core/blob/master/lib/transaction.js#L74)
- in [check bridge](https://github.com/leapdao/leap-node/blob/master/src/period/index.js), replace period submission with broadcast of periodSubmision txns
- in [check bridge](https://github.com/leapdao/leap-node/blob/master/src/period/index.js), build CAS bitmap and submit period if >2/3 of periodVotes available

## Deliverables
- code 
- tests

## Gain for the project
- ability to run PoS

## Roles
bounty gardener: name / share
bounty worker: name / share
bounty reviewer: name / share

### PR DESCRIPTION
**Pull request by troggy**
- [x] Validators emit PeriodVote tx when it is time to submit period.
- [x] Validators collect period votes in a state 

Resolves #282

### CODE REVIEW

File: src/period/submitPeriodVote.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    );
+    return;
+  }
+
+  const input = new Input(new Outpoint(periodRoot, 0));
+  const periodVoteTx = Tx.periodVote(mySlots[0].id, input).signAll(
+    bridgeState.account.privateKey
+  );
+
+  await sendDelayed(periodVoteTx);
```
Comment (johannbarbie) [inline src/period/submitPeriodVote.js]: why does this one need to be delayed? isn't tendermint deterministic? so if we got the period full, we are sure it will not go back?

Comment (troggy) [reply src/period/submitPeriodVote.js]: I have no answer to this yet, just used the existing approach: https://github.com/leapdao/leap-node/blob/master/src/eventsRelay.js#L65

File: src/tx/applyTx/checkPeriodVote.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    throw new Error(
+      `[period vote] Already submitted. Slot: ${slotId}. Root: ${periodRoot}`
+    );
+  }
+
+  state.periodVotes[periodRoot].push(slotId);
+
+  // validators are not in consensus on period roots atm
+  // disabling this until resolved
+  // https://github.com/leapdao/leap-node/issues/282#issuecomment-519160276
```
Comment (johannbarbie) [inline src/tx/applyTx/checkPeriodVote.js]: issue solved, should this be reenabled?

Comment (troggy) [reply src/tx/applyTx/checkPeriodVote.js]: it is still not working right. I'm fixing it as part of https://github.com/leapdao/leap-node/issues/309 here: https://github.com/leapdao/leap-node/tree/feat/submit-with-cas

TLDR: bad idea to submit period for tx handler. I'm submitting period for block middleware now, PR to follow once done

File: src/period/index.js [hunk 3]
```diff
@@ -19,8 +24,14 @@ module.exports = (bridgeState, nodeConfig = {}) => async (rsp, chainInfo, chainH
 
   logPeriod('checkBridge');
   // use a timeout to relax race conditions (if period submission is fast)
-  await new Promise((resolve) => {
+  await new Promise(resolve => {
     setTimeout(async () => {
+      await submitPeriodVote(
```
Comment (johannbarbie) [inline src/period/index.js]: why did you decide to call `submitPeriodVote()` here?

this is called when tendermint is trying to propose a new block (right?), and `submitPeriod()` is called as a fallback, in case the selected proposer is offline.

Comment (troggy) [reply src/period/index.js]: this handler is called every period (`checkPeriod` middleware). Right, I'm submitting the vote here as a fallback as well (`submitPeriodVote` has a protection against double votes).

`updatePeriod` is being called every block and that's where I submit a vote first: https://github.com/leapdao/leap-node/pull/305/files/15924f654aabe89a4492c03fa3c51f7f67b02545#diff-786c9ee7cb87a989994653f4a23150a6R25

File: src/txHelpers/submitPeriod.test.js [hunk 4]
```diff
@@ -180,6 +214,121 @@ describe('submitPeriod', () => {
     expect(submittedPeriod).toEqual({
       timestamp: '0',
     });
-    expect(bridgeState.operatorContract.test.submitCalled).toBe(false);
+    expect(submitPeriodWithCas).not.toBeCalled();
   });
+
+  // describe('period vote', () => {
```
Comment (johannbarbie) [inline src/txHelpers/submitPeriod.test.js]: no code comments

File: src/tx/applyTx/checkPeriodVote.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    !tx.inputs ||
+    !tx.inputs[0].signer ||
+    tx.inputs[0].signer !== state.slots[slotId].signerAddr
+  ) {
+    throw new Error(
+      `[period vote] Input should be signed by validator: ${state.slots[slotId].signerAddr}`
+    );
+  }
+
+  const periodRoot = bufferToHex(tx.inputs[0].prevout.hash);
```
Comment (johannbarbie) [inline src/tx/applyTx/checkPeriodVote.js]: also make sure that `tx.inputs[0].prevout.pos == 0`, otherwise it would screw up the index later.

File: src/tx/applyTx/checkPeriodVote.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+      `[period vote] Already submitted. Slot: ${slotId}. Root: ${periodRoot}`
+    );
+  }
+
+  state.periodVotes[periodRoot].push(slotId);
+
+  // validators are not in consensus on period roots atm
+  // disabling this until resolved
+  // https://github.com/leapdao/leap-node/issues/282#issuecomment-519160276
+  // const { result } = checkEnoughVotes(periodRoot, state);
```
Comment (johannbarbie) [inline src/tx/applyTx/checkPeriodVote.js]: remove code comments

**Comment by troggy** [comment]
addressed everything.

+ refactored to keep periodVotes in a tendermint `state`, not in a `bridgeState`. This is the right way to do ([see](https://leapdao.slack.com/archives/CAAGW85BR/p1565270194076200?thread_ts=1565264683.074200&cid=CAAGW85BR))

File: src/tx/applyTx/checkPeriodVote.test.js [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+  ...attrs,
+});
+
+const bridgeStateMock = attrs => ({
+  blockHeight: 1,
+  currentState: stateMock(),
+  ...attrs,
+});
+
+describe('checkPeriodVote', () => {
```
Comment (johannbarbie) [inline src/tx/applyTx/checkPeriodVote.test.js]: ❤️

File: src/period/submitPeriodVote.test.js [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+const PERIOD_ROOT =
+  '0x7777777777777777777777777777777777777777777777777777777777777777';
+const period = {
+  merkleRoot() {
+    return PERIOD_ROOT;
+  },
+  prevHash: '0x5678',
+};
+
+describe('submit period vote', () => {
```
Comment (johannbarbie) [inline src/period/submitPeriodVote.test.js]: ❤️
